### PR TITLE
refactor copy to api && change Reshape to lowercase && support more dtype && add more test

### PR DESCRIPTION
**Pull request by JiabinYang**

### CODE REVIEW

File: paddle/fluid/extension/include/dtype.h [hunk 1]
```diff
@@ -13,17 +13,26 @@ See the License for the specific language governing permissions and
...
+  COMPLEX128,
+  COMPLEX64,
+  FLOAT16,
   INT64,
   INT32,
+  INT16,
   UINT8,
   INT8,
-  // TODO(yangjiabin): Add other dtype support in next PR
+  // TODO(Superjomn) support more data types if needed.
```
Comment (chenwhql) [inline paddle/fluid/extension/include/dtype.h]: owner is you

File: paddle/fluid/extension/include/tensor.h [hunk 2]
```diff
@@ -57,25 +56,17 @@ class Tensor {
   /// It's usually used to set the input tensor data.
   /// \param data The pointer of the data, from which the tensor will copy.
```
Comment (chenwhql) [inline paddle/fluid/extension/include/tensor.h]: note also need to adjust

File: paddle/fluid/extension/src/tensor.cc [hunk 3]
```diff
@@ -12,10 +12,8 @@ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 See the License for the specific language governing permissions and
 limitations under the License. */
 
-#include "paddle/fluid/extension/include/tensor.h"
-
 #include <utility>
-
+#include "paddle/fluid/extension/include/all.h"
```
Comment (chenwhql) [inline paddle/fluid/extension/src/tensor.cc]: polish header format? style like:


image

https://zh-google-styleguide.readthedocs.io/en/latest/google-cpp-styleguide/headers/

File: paddle/fluid/framework/custom_operator.cc [hunk 4]
```diff
@@ -25,8 +25,7 @@ limitations under the License. */
 #include <utility>
 #include <vector>
 
-#include "paddle/fluid/extension/include/all.h"
-
+#include "paddle/fluid/extension/include/tensor.h"
```
Comment (chenwhql) [inline paddle/fluid/framework/custom_operator.cc:28]: I think all.h is better

File: paddle/fluid/framework/custom_operator.cc [hunk 5]
```diff
@@ -104,24 +103,61 @@ PlaceType ConvertInnerPlaceToEnumPlace(const platform::Place& pc) {
...
+      return proto::VarType::BF16;
+    case paddle::DataType::UINT8:
+      return proto::VarType::UINT8;
+    case paddle::DataType::INT8:
+      return proto::VarType::INT8;
+    case paddle::DataType::INT32:
+      return proto::VarType::INT32;
+    case paddle::DataType::INT64:
+      return proto::VarType::INT64;
     // TODO(chenweihang):
```
Comment (chenwhql) [inline paddle/fluid/framework/custom_operator.cc]: can be removed

File: paddle/fluid/framework/custom_operator.cc [hunk 6]
```diff
@@ -104,24 +103,61 @@ PlaceType ConvertInnerPlaceToEnumPlace(const platform::Place& pc) {
...
+    case proto::VarType::INT32:
+      return paddle::DataType::INT32;
+    case proto::VarType::INT8:
+      return paddle::DataType::INT8;
+    case proto::VarType::UINT8:
+      return paddle::DataType::UINT8;
+    case proto::VarType::INT16:
+      return paddle::DataType::INT16;
+
     // TODO(chenweihang):
```
Comment (chenwhql) [inline paddle/fluid/framework/custom_operator.cc]: same above

File: paddle/fluid/extension/src/tensor.cc [hunk 7]
```diff
@@ -85,129 +83,169 @@ DataType Tensor::type() const {
...
+    memory::Copy(gpu_place, static_cast<void *>(p_target_data), gpu_place,
+                 p_src_data, ele_size, dev_ctx->stream());
+  }
+  cudaStreamSynchronize(dev_ctx->stream());
+  return target;
 #else
-    PADDLE_THROW(platform::errors::Unavailable(
-        "Not compiled with CUDA, should not reach here."));
+  PADDLE_THROW(platform::errors::Unavailable(
+      "Not compiled with CUDA, should not reach here."));
```
Comment (chenwhql) [inline paddle/fluid/extension/src/tensor.cc]: polish error message, `Paddle is not compiled with CUDA.` is ok, `should not` is not needed

File: paddle/fluid/extension/src/tensor.cc [hunk 8]
```diff
@@ -85,129 +83,169 @@ DataType Tensor::type() const {
...
     auto *dev_ctx =
         static_cast<const platform::CUDADeviceContext *>(pool.Get(gpu_place));
-    memory::Copy(platform::CPUPlace(), static_cast<void *>(data), gpu_place,
-                 t_data, ele_num * sizeof(T), dev_ctx->stream());
+    memory::Copy(platform::CPUPlace(), static_cast<void *>(p_target_data),
+                 gpu_place, t_data, ele_num * sizeof(T), dev_ctx->stream());
 
     cudaStreamSynchronize(dev_ctx->stream());
 #else
     PADDLE_THROW(platform::errors::Unavailable(
```
Comment (chenwhql) [inline paddle/fluid/extension/src/tensor.cc:165]: same above

File: paddle/fluid/extension/src/tensor.cc (hunk 3 above)
Comment (JiabinYang) [reply paddle/fluid/extension/src/tensor.cc]: done

File: paddle/fluid/extension/src/tensor.cc (hunk 7 above)
Comment (JiabinYang) [reply paddle/fluid/extension/src/tensor.cc]: done

File: paddle/fluid/extension/src/tensor.cc (hunk 8 above)
Comment (JiabinYang) [reply paddle/fluid/extension/src/tensor.cc:165]: done

File: paddle/fluid/framework/custom_operator.cc (hunk 5 above)
Comment (JiabinYang) [reply paddle/fluid/framework/custom_operator.cc]: done

File: paddle/fluid/framework/custom_operator.cc (hunk 6 above)
Comment (JiabinYang) [reply paddle/fluid/framework/custom_operator.cc]: done

File: paddle/fluid/extension/include/tensor.h (hunk 2 above)
Comment (JiabinYang) [reply paddle/fluid/extension/include/tensor.h]: indent?

**Pull request merged**